crypto/param_build.c

```diff
@@ -277,25 +277,32 @@ static OSSL_PARAM *param_bld_convert(OSSL_PARAM_BLD *bld, OSSL_PARAM *param,
     return param;
 }
 
-OSSL_PARAM *ossl_param_bld_to_param(OSSL_PARAM_BLD *bld, void **secure)
+OSSL_PARAM *ossl_param_bld_to_param(OSSL_PARAM_BLD *bld, void **secure,
+                                    size_t *sec_sz)
 {
     OSSL_PARAM_BLD_BLOCK *blk, *s = NULL;
     OSSL_PARAM *param;
     const size_t p_blks = bytes_to_blocks((bld->curr + 1) * sizeof(*param));
     const size_t total = ALIGN_SIZE * (p_blks + bld->total_blocks);
+    size_t ss;
 
     if (bld->secure_blocks > 0) {
+        ss = bld->secure_blocks * ALIGN_SIZE;
+        if (sec_sz != NULL)
+            *sec_sz = ss;
         if (secure == NULL) {
```

Comment:
You should check `sec_sz` as well.

Reply:
?? I'm not sure what is being meant here.

I think I've addressed the review comments however.

Reply:
As in both _secure_ and _sec_sz_ have to be non-NULL or fail?
I guess I could.  It is ever meaningful for get _secure_ back without needing its size?  It's fine to pass to `OPENSSL_secure_free` without knowing the size.

Reply:
> As in both *secure* and *sec_sz* have to be non-NULL or fail?

Yes.  I see no reason why we shouldn't call `OPENSSL_secure_clear_free` (this is internal code, let's not be sloppy)

Reply:
I'd prefer to play it safe, but changing won't be difficult.
Plus, one of the test cases is (deliberately) sloppy :)

Reply:
Allowing internal sloppiness is playing it safe???

Reply:
Yes, check **all** arguments is playing it safe.  It isn't like any of this code will be time critical, we can afford the extra null check.

Reply:
I'm totally lost, or maybe you are.  In this thread, I'm arguing that you should check for `sec_sz` being NULL as well (and that being an error too), and you start with saying "no", but are now arguing that an extra NULL check is affordable.  Are we still talking about the same thing?

Reply:
Basically, I'm arguing that `secure` and `sec_sz` should both be non-NULL here, unconditionally.  They *may* both be NULL if no secure heap is allocated.

Reply:
I would also suggest that you add a ```free``` API and not have callers invoke two free routines because that ties you down, forever, to a single block which is an implementation detail.

Suggest is an understatement; I'm trying to be polite :)

Reply:
You don't want the contract of this API to be, forever and always, only allocate two contiguous blocks of memory.

Reply:
Note that this is an *internal* API, so changing it as needed at any time is allowed.

Reply:
I'm gong to do something along the lines of what @richsalz suggests.
It'll be cleaner for the caller.